asllib/ASTUtils.ml

```diff
@@ -593,60 +594,83 @@ let list_cross f li1 li2 =
     [] li1
   |> List.rev
 
+let list_flat_cross f li1 li2 =
+  List.fold_left
+    (fun xys x ->
+      List.fold_left (fun xys' y -> List.rev_append (f x y) xys') xys li2)
+    [] li1
+  |> List.rev
+
 let is_left_increasing = function
-  | MUL | DIV | DIVRM | SHL | SHR | POW | PLUS | MINUS -> true
-  | AND | BAND | BEQ | MOD | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MUL | SHL | SHR | POW | PLUS | MINUS -> true
+  | DIV | DIVRM | AND | BAND | BEQ | MOD | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_increasing = function
   | MUL | SHL | SHR | POW | PLUS -> true
-  | DIV | DIVRM | MINUS -> false
-  | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL | LT | LEQ
-  | NEQ | OR | RDIV ->
+  | MINUS -> false
+  | DIV | DIVRM | AND | BAND | MOD | BEQ | BOR | EOR | EQ_OP | GT | GEQ | IMPL
+  | LT | LEQ | NEQ | OR | RDIV ->
       assert false
 
 let is_right_decreasing = function
```

Comment:
(**
Tests whether the value of binary operation [op] decreases along with its right-hand-side operand when all operands are non-negative. That is, if [is_right_decreasing op] is true then for every two expressions [x op y1] and [x op y2] such that [y1 >= 0 /\ y2 >= 0 /\ x >=0] and [y1 <= y2] then [x op y1 <= x op y2].
*)